shark_turbine/kernel/wave/codegen.py

```diff
@@ -264,15 +264,27 @@ def handle_read(emitter: WaveEmitter, node: fx.Node):
     else:
         assert (
             mapping.is_output_identity()
-        ), "non-dentity output mapping is not supported yet"
-        input_mapping = mapping.map_input_indices(index.keys())
+        ), "non-identity output mapping is not supported yet"
+
+        # We need original symbolic shape to determine dimensions access order.
+        mem_shape = get_custom(memory).type.symbolic_shape
+
+        # Apply input order to input indices, e.g. if original input_mapping was
+        # {M: iter(0), N: iter(1)} and source mem has shape (N, M), result will
+        # be (iter(1), iter(0))
+        input_mapping = mapping.map_input_indices(mem_shape)
 
         iters = mapping.iters
+
+        # As we only support identity output mapping for now, we can directly
+        # substitute iterators with corresponding expanded index.
         subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
```

Comment:
Is there an assumption here about the ordering of the dimensions in index and iters? Say for example index was {K: ... , M: ... , N: ...} vs {M: .., N: ..., K: ...}, would that create a problem?

Reply:
I'm using mapping output shape for indexing if mapping is present, and for now, assert `mapping.is_output_identity()` is enforces output mapping is always `(iter(0), iter(1), ...)`